File: examples/cpp/galbot_robot/src/controller_management_example.cpp
#include <chrono>
#include <iostream>
#include <string>
#include <thread>
#include <vector>

#include "galbot_robot.hpp"

using namespace galbot::sdk::g1;

std::string controller_name_to_string(ControllerName name) {
  switch (name) {
    case ControllerName::CHASSIS_POSE_CTRL:
      return "CHASSIS_POSE_CTRL";
    case ControllerName::CHASSIS_TWIST_CTRL:
      return "CHASSIS_TWIST_CTRL";
    case ControllerName::LEG_PVT_BYPASS_CTRL:
      return "LEG_PVT_BYPASS_CTRL";
    case ControllerName::LEG_PVT_CTRL:
      return "LEG_PVT_CTRL";
    case ControllerName::HEAD_PVT_BYPASS_CTRL:
      return "HEAD_PVT_BYPASS_CTRL";
    case ControllerName::HEAD_PVT_CTRL:
      return "HEAD_PVT_CTRL";
    case ControllerName::LEFT_ARM_PVT_BYPASS_CTRL:
      return "LEFT_ARM_PVT_BYPASS_CTRL";
    case ControllerName::LEFT_ARM_PVT_CTRL:
      return "LEFT_ARM_PVT_CTRL";
    case ControllerName::RIGHT_ARM_PVT_BYPASS_CTRL:
      return "RIGHT_ARM_PVT_BYPASS_CTRL";
    case ControllerName::RIGHT_ARM_PVT_CTRL:
      return "RIGHT_ARM_PVT_CTRL";
    case ControllerName::LEFT_GRIPPER_CTRL:
      return "LEFT_GRIPPER_CTRL";
    case ControllerName::RIGHT_GRIPPER_CTRL:
      return "RIGHT_GRIPPER_CTRL";
    default:
      return "UNKNOWN";
  }
}

void print_active_controller(const std::string& group_name, ControllerName controller) {
  std::cout << "Active controller for " << group_name << ": " << controller_name_to_string(controller) << std::endl;
}

// Helper function to print control status
void print_status(const std::string& operation, ControlStatus status) {
  std::cout << operation << ": ";
  if (status == ControlStatus::SUCCESS) {
    std::cout << "SUCCESS";
  } else {
    std::cout << "FAILED (Code: " << static_cast<int>(status) << ")";
  }
  std::cout << std::endl;
}

int main() {
  // Get robot instance
  auto& robot = GalbotRobot::get_instance();

  // Initialize system
  std::cout << "Initializing robot..." << std::endl;
  if (robot.init()) {
    std::cout << "System initialized successfully!" << std::endl;
  } else {
    std::cerr << "System initialization failed!" << std::endl;
    return -1;
  }

  // Wait for data readiness
  std::this_thread::sleep_for(std::chrono::seconds(1));

  // Define the controller we want to test
  // Using LEFT_ARM_PVT_CTRL as the test subject
  ControllerName test_controller = ControllerName::LEFT_ARM_PVT_CTRL;
  JointGroup test_group = JointGroup::LEFT_ARM;
  std::string controller_str = "LEFT_ARM_PVT_CTRL";
  std::string group_str = "LEFT_ARM";

  std::cout << "\n--- Starting Controller Management Example ---\n" << std::endl;

  // 1. Get active controller before switching
  ControllerName active_controller = robot.get_active_controller(test_group);
  print_active_controller(group_str, active_controller);

  // 2. Switch controller
  // Demonstrates switch_controller interface
  // This ensures the controller is acquired and started
  std::cout << "\n[Test 1] Switching to " << controller_str << "..." << std::endl;
  ControlStatus status = robot.switch_controller(test_controller);
  print_status("switch_controller", status);
  
  std::this_thread::sleep_for(std::chrono::milliseconds(500));
  active_controller = robot.get_active_controller(test_group);
  print_active_controller(group_str, active_controller);

  // 3. Stop controller
  // Demonstrates stop_controller interface (JointGroup overload)
  std::cout << "\n[Test 2] Stopping controller for group " << group_str << "..." << std::endl;
  status = robot.stop_controller(test_group);
  print_status("stop_controller", status);

  std::this_thread::sleep_for(std::chrono::milliseconds(500));

  // 4. Release controller
  // Demonstrates release_controller interface (JointGroup overload)
  std::cout << "\n[Test 3] Releasing controller for group " << group_str << "..." << std::endl;
  status = robot.release_controller(test_group);
  print_status("release_controller", status);

  std::this_thread::sleep_for(std::chrono::milliseconds(500));

  // 5. Acquire controller
  // Demonstrates acquire_controller interface
  std::cout << "\n[Test 4] Acquiring " << controller_str << "..." << std::endl;
  status = robot.acquire_controller(test_controller);
  print_status("acquire_controller", status);

  std::this_thread::sleep_for(std::chrono::milliseconds(500));

  // 6. Start controller
  // Demonstrates start_controller interface (JointGroup overload)
  std::cout << "\n[Test 5] Starting controller for group " << group_str << "..." << std::endl;
  status = robot.start_controller(test_group);
  print_status("start_controller", status);

  std::this_thread::sleep_for(std::chrono::milliseconds(500));

  std::cout << "\n--- Controller Management Example Finished ---\n" << std::endl;

  // Shutdown
  robot.request_shutdown();
  robot.wait_for_shutdown();
  robot.destroy();

  return 0;
}

File: examples/python/galbot_robot/controller_management_example.py
import time
import time

from galbot_sdk.g1 import GalbotRobot, ControlStatus, ControllerName, JointGroup

def controller_name_to_string(name):
    mapping = {
        ControllerName.CHASSIS_POSE_CTRL: "CHASSIS_POSE_CTRL",
        ControllerName.CHASSIS_TWIST_CTRL: "CHASSIS_TWIST_CTRL",
        ControllerName.LEG_PVT_BYPASS_CTRL: "LEG_PVT_BYPASS_CTRL",
        ControllerName.LEG_PVT_CTRL: "LEG_PVT_CTRL",
        ControllerName.HEAD_PVT_BYPASS_CTRL: "HEAD_PVT_BYPASS_CTRL",
        ControllerName.HEAD_PVT_CTRL: "HEAD_PVT_CTRL",
        ControllerName.LEFT_ARM_PVT_BYPASS_CTRL: "LEFT_ARM_PVT_BYPASS_CTRL",
        ControllerName.LEFT_ARM_PVT_CTRL: "LEFT_ARM_PVT_CTRL",
        ControllerName.RIGHT_ARM_PVT_BYPASS_CTRL: "RIGHT_ARM_PVT_BYPASS_CTRL",
        ControllerName.RIGHT_ARM_PVT_CTRL: "RIGHT_ARM_PVT_CTRL",
        ControllerName.LEFT_GRIPPER_CTRL: "LEFT_GRIPPER_CTRL",
        ControllerName.RIGHT_GRIPPER_CTRL: "RIGHT_GRIPPER_CTRL",
    }
    return mapping.get(name, "UNKNOWN")


def print_active_controller(group_name, controller_name):
    print(f"Active controller for {group_name}: {controller_name_to_string(controller_name)}")

def print_status(operation, status):
    if status == ControlStatus.SUCCESS:
        print(f"{operation}: SUCCESS")
    else:
        # Assuming status can be cast to int or printed directly
        print(f"{operation}: FAILED (Status: {status})")

def main():
    # Get robot instance and init
    robot = GalbotRobot.get_instance()
    print("Initializing robot...")
    if robot.init():
        print("System initialized successfully!")
    else:
        print("System initialization failed!")
        return

    # Wait for data readiness
    time.sleep(1)

    # Define the controller we want to test
    # Using LEFT_ARM_PVT_CTRL as the test subject
    test_controller = ControllerName.LEFT_ARM_PVT_CTRL
    controller_str = "LEFT_ARM_PVT_CTRL"
    # Using string for display and JointGroup enum for querying
    group_str = "left_arm"
    test_group = JointGroup.LEFT_ARM

    print("\n--- Starting Controller Management Example ---")

    # 1. Get active controller before switching
    active_controller = robot.get_active_controller(test_group)
    print_active_controller(group_str, active_controller)

    # 2. Switch controller
    print(f"\n[Test 1] Switching to {controller_str}...")
    status = robot.switch_controller(test_controller)
    print_status("switch_controller", status)
    
    time.sleep(0.5)
    active_controller = robot.get_active_controller(test_group)
    print_active_controller(group_str, active_controller)

    # 3. Stop controller
    print(f"\n[Test 2] Stopping controller for group {group_str}...")
    # Using string overload/version for group
    status = robot.stop_controller(group_str)
    print_status("stop_controller", status)

    time.sleep(0.5)

    # 4. Release controller
    print(f"\n[Test 3] Releasing controller for group {group_str}...")
    status = robot.release_controller(group_str)
    print_status("release_controller", status)

    time.sleep(0.5)

    # 5. Acquire controller
    print(f"\n[Test 4] Acquiring {controller_str}...")
    status = robot.acquire_controller(test_controller)
    print_status("acquire_controller", status)

    time.sleep(0.5)

    # 6. Start controller
    print(f"\n[Test 5] Starting controller for group {group_str}...")
    status = robot.start_controller(group_str)
    print_status("start_controller", status)

    time.sleep(0.5)

    print("\n--- Controller Management Example Finished ---")

    # Shutdown
    robot.request_shutdown()
    robot.wait_for_shutdown()
    robot.destroy()
    print("Resources released successfully")

if __name__ == "__main__":
    main()
